### ISSUE TITLE
Training on energies, forces, and stress?

### ISSUE THREAD

**Issue by anywallsocket** [atomistic-machine-learning/schnetpack#551]
**I can train on forces and energies as per the tutorial, but I'm having trouble implementing the stress calculation.**

My config is basically identical to the one exemplified by the force fields tutorial seen here: https://schnetpack.readthedocs.io/en/latest/tutorials/tutorial_05_materials.html

I followed the instructions for adding stress as such,

Add stress key and calc_stress = True tag for Forces output module:
```
pred_energy = spk.atomistic.Atomwise(n_in=n_atom_basis, output_key='energy')
pred_forces = spk.atomistic.Forces(energy_key='energy', force_key='forces', stress_key='stress', calc_stress=True)
```

Add stress predictions as output for loss calculation in the model:
```
output_stress = spk.task.ModelOutput(
    name='stress',
    loss_fn=torch.nn.MSELoss(),
    loss_weight=100.,
    metrics={"MAE": torchmetrics.MeanAbsoluteError()}
)
```

Add Strain module to input module within the model: (do I need pred_stress here?)
```
input_modules=[spk.atomistic.Strain(), spk.atomistic.PairwiseDistances()],
output_modules=[pred_energy, pred_forces] 
```

Which are connected to the atomistic task:

`outputs=[output_energy, output_forces, output_stress],
`

The first warning I received is:
_One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior_

If this is mitigated, I get batch size / tensor size errors:

This is after I realized SchNetPack expect stress to be voigt=False unlike ASE default.
So my stress shape is `torch.Size([3,3])` as I believe it should be.

With batch size e.g.,  = 10 I get:
_The size of tensor a (3) must match the size of tensor b (30) at non-singleton dimension 1_

Yet batch size = 1 I get:
_Predictions and targets are expected to have the same shape, but got torch.Size([1, 3, 3]) and torch.Size([3, 3])_

**I believe I am close but have spent too much time on it, so please let me know how to properly implement stress.**

Additionally, if I run your force field tutorial as is (link above), I relieve the value error:

_can only convert an array of size 1 to a Python scalar_

for the following line (which I have checked the property, and it is trying to unpack stress not energy):

```
File ~/miniconda3/lib/python3.9/site-packages/schnetpack/interfaces/ase_interface.py:305, in SpkCalculator.calculate(self, atoms, properties, system_changes)
    301 if model_prop in model_results:
    302     if prop == self.energy or prop == self.stress:
    303         # ase calculator should return scalar energy
    304         results[prop] = (
--> 305             model_results[model_prop].cpu().data.numpy().item()
    306             * self.property_units[prop]
```

I figured these two issues might be related?

**Thank you very much, any and all help will be greatly appreciated!**

**Comment by anywallsocket** [comment]
Thank you all, I have figured it out.

One thing to note is that the initial error running the Force Fields tutorial as-is, still persists.
This is solved just by distinguishing energies and forces in the line

` 302     if prop == self.energy or prop == self.stress:`

Because it seems you cannot `.item()` stress, as its shape is `torch.Size([3,3])` or `torch.Size([1,3,3])`, the latter of which seems expected by default, meaning one needs to bracket stress properties upon loading into the datamodule, and likewise use `model_results[model_prop].cpu().data.numpy()[0]` for stress. 

Perhaps I've got this wrong, but it's how I've managed to fix the aforementioned issues.

Note as well that the warning

> One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior

should _NOT_ be mitigated in `torch/autograd/__init__.py` as the fix uses all tensors as it should.

Best,
- any
 
I'll leave this here just in case anyone else has related issues for now.
